Add Github organization collaborators fetcher

mlomena opened the pull request:
- [x] Tick to sign-off your agreement to the [Developer Certificate of Origin (DCO) 1.1](../blob/master/DCO1.1.txt)


## What

Adds fetcher to retrieve collaborators of repositories in a Github organization

## Why

_Provide the business justification for the work included in the pull request._

## How

- Add fetcher `fetch_org_collaborators.py`
- Add evidence class in `gh_org_collaborators.py`
- Updated `devel.json` and `README.md`

## Context

- redacted

drsm79 commented:
I think this should be under a GitHub provider.  The auditree fetchers/checks are verifying the correct configuration of the tool, which this isn’t really related to.

alfinkel commented:
@drsm79 I understand the logic behind your [comment about this going under a Github provider](https://github.com/ComplianceAsCode/auditree-arboretum/pull/38#issuecomment-719449293) folder and if this were just a fetcher thing then I think it could make sense.  However, I think that any checks written against this sort of evidence could possibly be applied to Bitbucket and GitLab repo hosting services too.  Ideally this would go under some provider and within that provider we'd have fetchers in github, bitbucket, and gitlab subfolders and checks in their checks folder.  I figured that since the auditree provider is already structured that way for repo integrity f/c's we could continue to put things in there.  But I see your point.  I suppose we could come up with another generic "provider" to put this under and then follow the same pattern as we have for repo integrity.  What do you think about `repo_host_service` or `repo_host` or `rhs` as the provider?

alfinkel reviewed (changes requested):
As discussed over Slack, rather than using PyGithub you should use the Framework Github Service which now has the `make_request` and `paginate_api` methods exposed as of v1.6.5 so you can make direct requests to the GH API in cases where the service is lacking in functionality support.

Also, for your evidence we would need unit tests included as well.

alfinkel commented:
@mlomena is there a check in the works?

alfinkel commented:
re: [fetcher location](https://github.com/ComplianceAsCode/auditree-arboretum/pull/38#issuecomment-719449293) - I've added a "permissions" vendor in PR #39 and this github fetcher should live under there.  Can you move all of the content in this PR to the appropriate sub-folders under the "permissions" vendor?

mstancamp reviewed (commented):
With PR #39 should all the existing fetchers under auditree/fetchers/github and the new github fetchers be placed under the new permissions vendor? Will the fetchers be under the fetchers subdirectory or under a categorization subdirectory like permissions/fetchers/github?

alfinkel commented:
The "auditree" vendor will stay as is.  The fetcher in this PR should go under the "permissions" vendor under "fetchers" and the "github" technology.  So following the pattern in the auditree vendor structure: `arboretum/permissions/fetchers/github`.

mstancamp reviewed (commented):
What is the status of this PR? We have the following support that will be added once this PR approved and merged:
- Three additional fetch methods to retrieve evidence for teams, team members. forks. 
- A check to report on the status of the repos based on the evidence collected. We will be reviewing the check report with the issue creator this week.

alfinkel commented:
I believe we're waiting on unit tests for the evidence code: See: https://github.com/ComplianceAsCode/auditree-arboretum/pull/38#pullrequestreview-521848231 - specifically:

> ...
> Also, for your evidence we would need unit tests included as well.

mlomena commented:
@alfinkel sorry for the delay but I am now back from a long absence.
I have moved fetcher, evidence and the content from README to the permissions folder.
I have also added unit tests.

mlomena commented:
@alfinkel Sorry for the long delay. I am now back from a long absence.
I have moved the stuff in this PR to `permissions` and have added unit tests.

alfinkel commented on arboretum/permissions/README.md:
```diff
@@ -16,7 +16,73 @@ how to include the fetchers and checks from this library in your downstream proj
-Fetchers coming soon...
+### Organization Integrity (Repository Collaborators)
+
+* Class: [GithubOrgCollaboratorsFetcher](https://github.com/ComplianceAsCode/auditree-arboretum/blob/main/arboretum/permissions/fetchers/github/fetch_org_collaborators.py)
```
Put the link down below with the other links to match readme links pattern.

mlomena replied:
The proposed change was committed.

alfinkel commented on arboretum/permissions/README.md:
```diff
@@ -16,7 +16,73 @@ how to include the fetchers and checks from this library in your downstream proj
+* Behavior: For each Github organization specified, an evidence file is stored in the locker containing the collaborators details. If no repositories are specified the fetcher defaults to retrieving the collaborators in all the repositories in the organization. TTL is set to 1 day.
+* Configuration elements:
+  * `org.permissions.org_integrity.orgs`
+     * Optional
```
`org.permissions.org_integrity.orgs` isn't optional.

mlomena replied:
The proposed change was committed.

alfinkel commented on arboretum/permissions/README.md:
```diff
@@ -16,7 +16,73 @@ how to include the fetchers and checks from this library in your downstream proj
+        * Key: `repos` (string) (optional)
+        * Value: List of repository names (string) for the Github organization
+        * Key: `collaborator_types` (string)
+        * Value: List of collaborator types (string) (Options are `all`, `direct` and `outside`)
```
Please follow the documentation pattern used [here](https://github.com/ComplianceAsCode/auditree-arboretum/tree/main/arboretum/issue_mgmt#github-issues) and [here](https://github.com/ComplianceAsCode/auditree-arboretum/tree/main/arboretum/issue_mgmt#zenhub-workspaces).

alfinkel replied:
It seems as though `url` should be required and `repos` and `collaborator_types` should be optional.  Shouldn't the default be "all" and the optional values be limited to "direct" and "outside"?

alfinkel commented on arboretum/permissions/README.md:
```diff
@@ -16,7 +16,73 @@ how to include the fetchers and checks from this library in your downstream proj
+        * Value: List of repository names (string) for the Github organization
+        * Key: `collaborator_types` (string)
+        * Value: List of collaborator types (string) (Options are `all`, `direct` and `outside`)
+* Example (optional) configuration:
```
Not optional.

```suggestion
* Example configuration:
```

alfinkel commented on arboretum/permissions/README.md:
````diff
@@ -16,7 +16,73 @@ how to include the fetchers and checks from this library in your downstream proj
+  ```
+
+* Required credentials:
+  * `github` or `github_enterprise` credentials with permissions to the repositories are required for this fetcher to successfully retrieve evidence.
````
`...credentials with permissions to the repositories...` - You should be more clear about what kind of permissions are required for this fetcher to run successfully.

alfinkel replied:
This still needs to be addressed.

mlomena replied:
Fixed with a link to github repository permissions

alfinkel commented on arboretum/permissions/evidences/github/gh_org_collaborators.py:
```diff
@@ -0,0 +1,64 @@
+# -*- mode:python; coding:utf-8 -*-
```
I don't think this needs its own `github` folder under `permissions/evidences`.  Suggest removing the `github` folder and storing this module directly under `arboretum/permissions/evidences`.

alfinkel commented on arboretum/permissions/evidences/github/gh_org_collaborators.py:
```diff
@@ -0,0 +1,64 @@
+        if self.content:
+            if not hasattr(self, '_as_a_dict'):
+                self._as_a_dict = json.loads(self.content)
+            return self._as_a_dict
```
This isn't needed.  Use the evidence [content_as_json](https://github.com/ComplianceAsCode/auditree-framework/blob/main/compliance/evidence.py#L113) property instead.

alfinkel commented on devel.json:
```diff
@@ -23,6 +23,20 @@
+            "url": "https://github.my-company.com/my-org-2",
+            "collaborator_types": ["direct", "outside"]
+          }
+        ]
```
This json is invalid.  I would think a test would have caught that?

mlomena replied:
This was fixed and committed

alfinkel commented on arboretum/permissions/fetchers/github/fetch_org_collaborators.py:
```diff
@@ -0,0 +1,60 @@
+from compliance.utils.services.github import Github
+
+
+class GithubOrgCollaboratorsFetcher(ComplianceFetcher):
```
This fetcher should be refactored to use the [raw_evidence context manager](https://github.com/ComplianceAsCode/auditree-framework/blob/main/compliance/evidence.py#L283-L308).  You should also follow the pattern of adding the evidence to the evidence cache rather than how you're handling it right now.  See [here](https://github.com/ComplianceAsCode/auditree-arboretum/blob/main/arboretum/issue_mgmt/fetchers/fetch_github_issues.py#L37-L63).

See doc [ref](https://complianceascode.github.io/auditree-framework/design-principles.html#evidence-validation).

alfinkel reviewed (changes requested):
As I state in a review comment, I think your fetcher needs refactoring to leverage more of the framework.  In addition to that though, the relationship between the collaborator types is unclear as is the way you're storing the evidence.  Do you intend to do something with the information of whether a collaborator is direct or outside later on in a check?  It seems that "all" collaborators would equal to "direct" + "outside" collaborators?  What's the reason for even having a collaborator_types option?  My suggestion here is to confirm whether "all" is the same as "direct" + "outside" and then based on whether you actually want to track a collaborator's type (affiliation) make either a single call to get "all" collaborators or make two separate calls to get "direct" and "outside" collaborators.  If you go the latter route then I suggest composing the evidence by repo and then affiliation rather than the other way around as you have it now.

Also there doesn't appear to be any use for your subclassed evidence class _at the moment_ unless you have a check coming?  If you do then I suggest removing the evidence class (and tests) from this PR and including it with any forthcoming checks PR.  It would probably make this review go smoother.

alfinkel commented on arboretum/permissions/README.md:
```diff
@@ -16,7 +16,73 @@ how to include the fetchers and checks from this library in your downstream proj
+        * Key: `repos` (string) (optional)
+        * Value: List of repository names (string) for the Github organization
+        * Key: `collaborator_types` (string) (optional)
+        * Value: List of collaborator types (string) (Options are `all`, `direct` and `outside`. Default is `all`.)
```
Looks like you've not applied [this earlier comment](https://github.com/ComplianceAsCode/auditree-arboretum/pull/38#discussion_r551936699).

```suggestion
  * `org.permissions.org_integrity.orgs`
     * Required
     * List of dictionaries each containing Github organization retrieval configuration.
        * `url`
           * Required
           * String in the form of `https://github.com/my-org`.
           * Use to define the Github organization where collaborator information will be extracted.
        * `repos`
           * Optional
           * List of strings in the form of `["my-repo", "my-other-repo"]`.
           * Defaults to all repositories in the organization.
           * Use if looking to filter collaborator evidence to a subset of repositories in the organization otherwise do not include.
        * `collaborator_types`
           * Optional
           * List of strings in the form of `["all", "direct", "outside"]`.
           * Valid list element values are `"all"`, `"direct"`, `"outside"`.
           * Defaults to `["all"]`.
           * Use if looking to override retrieval of all collaborators otherwise do not include.
```

mlomena replied:
The proposed change was committed

alfinkel commented on arboretum/permissions/fetchers/github/fetch_org_collaborators.py:
```diff
@@ -0,0 +1,59 @@
+            parsed = urlparse(org['url'])
+            gh_org = parsed.path.strip('/')
+            file_prefix = gh_org.lower().replace('-', '_')
+            base_url = f'{parsed.scheme}://{parsed.hostname}'
```
This all seems unnecessary.  Again, I suggestion you reference [this fetcher](https://github.com/ComplianceAsCode/auditree-arboretum/blob/main/arboretum/issue_mgmt/fetchers/fetch_github_issues.py#L37-L63) that pretty much does the same thing.  Rather than including the organization name as part of the evidence name, it uses a 10 char hash based on config parameters supplied by a [utility function found in the framework](https://github.com/ComplianceAsCode/auditree-framework/blob/main/compliance/utils/data_parse.py#L40).  So ideally you'd have evidence files named similar to "gh_org_all_collaborators_1ab2c3d4e5.json".  Any check can then use the same function to target the desired evidence based on the same config that was used by the fetcher.

mlomena replied:
Changed to use hash

alfinkel commented on arboretum/permissions/fetchers/github/fetch_org_collaborators.py:
```diff
@@ -0,0 +1,59 @@
+            gh_org = parsed.path.strip('/')
+            file_prefix = gh_org.lower().replace('-', '_')
+            base_url = f'{parsed.scheme}://{parsed.hostname}'
+            github = Github(base_url=base_url)
```
It's inefficient to create a new Github object for every call you make.  It defeats part of the purpose of having a Github object which is having an open session that can be reused without having to authenticate for each call.  Instead, again, I point you to [the Github issues fetcher](https://github.com/ComplianceAsCode/auditree-arboretum/blob/main/arboretum/issue_mgmt/fetchers/fetch_github_issues.py) that uses a pool of reusable Github objects.

mlomena replied:
This was discussed and fixed

alfinkel commented on arboretum/permissions/fetchers/github/fetch_org_collaborators.py:
```diff
@@ -0,0 +1,59 @@
+            affiliations = (
+                org['collaborator_types']
+                if 'collaborator_types' in org else ['all']
+            )
```
"all" should be a constant added to the common.constants module (with a comment describing its purpose).  This is something we agreed to with the auditors.  Suggest:

```suggestion
            affiliations = org.get('collaborator_types', ALL_COLLABORATORS)
```

alfinkel commented on arboretum/permissions/fetchers/github/fetch_org_collaborators.py:
```diff
@@ -0,0 +1,59 @@
+                org['collaborator_types']
+                if 'collaborator_types' in org else ['all']
+            )
+            for a in affiliations:
```
```suggestion
            for affiliation in affiliations:
```

alfinkel commented on devel.json:
```diff
@@ -25,6 +25,20 @@
+        "orgs": [
+          {
+            "url": "https://github.com/my-org-1"
+	  },
```
```suggestion
	    },
```

alfinkel reviewed (changes requested):
It looks like you still need to address a couple of comments from the previous review.

- https://github.com/ComplianceAsCode/auditree-arboretum/pull/38#discussion_r551940617
- https://github.com/ComplianceAsCode/auditree-arboretum/pull/38#discussion_r551936699
- You should also consider just copying the [GH issues fetcher](https://github.com/ComplianceAsCode/auditree-arboretum/blob/main/arboretum/issue_mgmt/fetchers/fetch_github_issues.py) and then updating it to suit your needs.  It does something very similar and handles reusing Github service objects for GH and GHE as well as naming evidence based on a sha256 hash composed by using unique config values as the seed.
- You also need an update to the CHANGES.md

alfinkel commented on arboretum/permissions/README.md:
```diff
@@ -16,7 +16,87 @@ how to include the fetchers and checks from this library in your downstream proj
+
+* Class: [GithubOrgCollaboratorsFetcher][gh-org-fetcher]
+* Purpose: Writes the details of collaborators in Github organizations to the evidence locker. This fetcher class is only meant for use with Github or Github Enterprise organizations.
+* Behavior: For each Github organization specified, an evidence file per collaborator type is stored in the locker containing the collaborators details of all the repositories in the organization. If no repositories are specified the fetcher defaults to retrieving the collaborators in all the repositories in the organization. If no collaborator types are specified the fetcher defaults to retrievi ...
```
```suggestion
* Behavior: For each Github organization specified, an evidence file per collaborator type (affiliation) is stored in 
the locker containing collaborator details for the specified repositories in the organization.  The default is to 
retrieve all collaborators by affiliation from all repositories in each specified Github organization.  TTL is set to 1 
day.
```

mlomena replied:
The proposed change has been committed.

alfinkel commented on arboretum/permissions/README.md:
```diff
@@ -16,7 +16,87 @@ how to include the fetchers and checks from this library in your downstream proj
+  * `org.permissions.org_integrity.orgs`
+     * Required
+     * List of dictionaries each containing Github organization retrieval configuration.
+        * `name`
+           * Required
+           * String in the form of `my-org`.
+           * Use to define the organization to retrieve the collaborators.
+        * `host`
+           * Optional
+           * String in the form of `https://github.<company>.com` for Enterprise GitHub..
+           * Defaults to `https://github.com`.
+           * Use to define the Github base url to use.
```
There's no need to split this up into name and host.  You should return this to just the `url` required field and then in the code you would separate whether it's GH or GHE.  Plus at least then it will be in sync with you example below which is still using `url`.

mlomena replied:
This was done following the same approach in your [issues fetcher](https://github.com/ComplianceAsCode/auditree-arboretum/blob/main/arboretum/issue_mgmt/fetchers/fetch_github_issues.py#L40-L41).
We can discuss it during the call.
Agreed that the example does not match the implementation.

mlomena replied:
Discussed and fixed to use `url` only

alfinkel commented on arboretum/permissions/README.md:
````diff
@@ -16,7 +16,87 @@ how to include the fetchers and checks from this library in your downstream proj
+  ```
+
+* Required credentials:
+  * `github` or `github_enterprise` credentials with [appropriate permissions to the repositories](https://docs.github.com/en/free-pro-team@latest/github/setting-up-and-managing-organizations-and-teams/repository-permission-levels-for-an-organization) are required for this fetcher to successfully retrieve evidence.
````
Please put this link at the bottom of the page as we do with the rest of the links.

mlomena replied:
Fixed and committed

alfinkel commented on arboretum/permissions/fetchers/github/fetch_org_collaborators.py:
```diff
@@ -0,0 +1,70 @@
+# -*- mode:python; coding:utf-8 -*-
+# Copyright (c) 2020 IBM Corp. All rights reserved.
```
```suggestion
# Copyright (c) 2021 IBM Corp. All rights reserved.
```

mlomena replied:
Fixed

alfinkel commented on arboretum/permissions/fetchers/github/fetch_org_collaborators.py:
```diff
@@ -0,0 +1,70 @@
+    def fetch_gh_org_collaborators(self):
+        """Fetch collaborators from GH organization repositories."""
+        for org in self.orgs:
+            host = org.get('host', GH_HOST_URL)
+            if host not in self.gh_pool.keys():
+                self.gh_pool[host] = Github(base_url=host)
+            repos = self.gh_pool[host].paginate_api(
+                f'orgs/{org["name"]}/repos'
+            )
```
You shouldn't do this until you're inside of the `raw_evidence` context manager otherwise it could be doing unnecessary work.  Calls to get repos from the API can be expensive depending on how many repos there are so you don't want to do them unless you have to.

mlomena replied:
If done inside the `raw_evidence` we would need a call to get the repos on each `affiliation` type.
We can discuss during the call.

mlomena replied:
Agreed. We can store the repos in a dictionary so they do not have to be retrieved for each affiliation type.

mlomena replied:
Discussed and fixed

alfinkel commented on arboretum/permissions/fetchers/github/fetch_org_collaborators.py:
```diff
@@ -0,0 +1,70 @@
+                for repo in repos:
+                    if 'repos' in org and repo['name'] not in org['repos']:
+                        continue
+                    users = self.gh_pool[host].paginate_api(
+                        f'repos/{org["name"]}/{repo["name"]}/collaborators',
+                        affiliation=affiliation
+                    )
```
Same here, this logic should all go inside the `if` block in the context manager.  These API calls are expensive so we only want to do them when it is necessary.

mlomena replied:
Agreed.

mlomena replied:
Discussed and fixed

alfinkel commented on CHANGES.md:
```diff
@@ -1,3 +1,7 @@
+# [0.8.0](https://github.com/ComplianceAsCode/auditree-arboretum/releases/tag/v0.8.0)
+
+- [ADDED] GitHub org collaborators fetcher added to Permissions.
```
```suggestion
- [ADDED] GitHub org collaborators fetcher added to `Permissions`.
```

mlomena replied:
Fixed.

alfinkel commented on devel.json:
```diff
@@ -25,6 +25,21 @@
     },
+    "permissions": {
+      "org_integrity": {
+        "orgs": [
+          {
+            "name": "my-org-1"
+          },
+          {
+            "name": "my-org-2",
+            "host": "https://github.my-company.com",
+            "collaborator_types": ["direct", "outside"],
+            "repos": ["repo1", "repo2"]
+          }
+        ]
```
This should be changed back to using `url` rather than name and host.

mlomena replied:
To discuss. Same as per https://github.com/ComplianceAsCode/auditree-arboretum/pull/38#discussion_r559445782.

mlomena replied:
Fixed after discussing

alfinkel reviewed (changes requested):
You're currently not leveraging the framework to optimize your fetcher.  I've set up a call for Monday to clear things up.  In the meantime can you update the docs and the copyright notice dates?  Thanks.

alfinkel commented on arboretum/permissions/README.md:
```diff
@@ -16,7 +16,84 @@ how to include the fetchers and checks from this library in your downstream proj
+
+* Class: [GithubOrgCollaboratorsFetcher][gh-org-fetcher]
+* Purpose: Writes the details of collaborators in Github organizations to the evidence locker. This fetcher class is only meant for use with Github or Github Enterprise organizations.
+* Behaviour: For each Github organization specified, an evidence file per collaborator type (affiliation) is stored in
```
```suggestion
* Behavior: For each Github organization specified, an evidence file per collaborator type (affiliation) is stored in
```

To match the rest of the repo.

alfinkel commented on arboretum/permissions/fetchers/github/fetch_org_collaborators.py:
```diff
@@ -0,0 +1,65 @@
+            host, org = config['url'].rsplit('/', 1)
+            for aff in config.get('collaborator_types', GH_ALL_COLLABORATORS):
+                url_hash = get_sha256_hash([config['url']], 10)
+                json_file = (f'gh_{aff}_collaborators_{url_hash}.json')
```
```suggestion
                json_file = f'gh_{aff}_collaborators_{url_hash}.json'
```

alfinkel commented on arboretum/permissions/README.md:
```diff
@@ -16,7 +16,84 @@ how to include the fetchers and checks from this library in your downstream proj
+    "org": {
+      "permissions": {
+        "org_integrity": {
+          "orgs": [
+            {
+              "url": "https://github.com/my-org-1",
+              "repos": ["repo1", "repo2"],
+            },
+            {
+              "url": "https://github.my-company.com/my-org-2",
+              "collaborator_types": ["direct", "outside"]
+            }
+          ]
```
```suggestion
          "orgs": [
            {
              "url": "https://github.com/my-org-1"
            },
            {
              "url": "https://github.my-company.com/my-org-2",
              "collaborator_types": ["direct", "outside"],
              "repos": ["repo1", "repo2"]
            }
          ]
```

To match the devel.json "example".

alfinkel commented on arboretum/permissions/README.md:
```diff
@@ -16,7 +16,84 @@ how to include the fetchers and checks from this library in your downstream proj
+     * List of dictionaries each containing Github organization retrieval configuration.
+        * `url`
+           * Required
+           * String in the form of `https://github.<company>.com/my-org`.
```
```suggestion
           * String in the form of `"https://github.com/my-org"` or `"https://github.<company>.com/my-org"`.
```

alfinkel reviewed (changes requested):
Just some minor things...

alfinkel commented:
Can you squash commits down to one signed commit?

mlomena committed:
Add GitHub repo collaborators fetcher

mlomena committed:
Delete DS_Store files

mlomena committed:
Delete more .DS_Store files

alfinkel reviewed (approved):
+1
